### PR TITLE
feat: Add support for TypeScript tuple types

### PR DESCRIPTION
**Pull request by ExE-Boss**
Part of #50

This implements parsing and handling of TypeScript’s tuple types, which are rather common.

Note that this will have merge conflicts with #60.

---

review?(@sandersn)

### CODE REVIEW

File: tests/fixtures/typescript-types [hunk 1]
```diff
@@ -28,7 +28,7 @@
 {function(() => string): void}
 {X<() => string>}
 {new (x: number) => string}
-//{[it, needs, to, be]}
+{[it, needs, to, be]}
```
Comment (sandersn) [inline tests/fixtures/typescript-types]: can you add a few more tests? I'd like to see one per rule that got altered in the pegjs file, although that's hard to verify sometimes.

Also tests of `[]`, `[x]` and `[][]`.

File: peg_src/jsdoctype.pegjs [hunk 2]
```diff
@@ -907,6 +917,45 @@ RecordTypeExprEntryOperand = UnionTypeExpr
                            / UnknownTypeExpr
 
 
+/**
+ * TypeScript style tuple type.
+ *
+ * Examples:
+ *   - [string]
```
Comment (sandersn) [inline peg_src/jsdoctype.pegjs]: this parses `[]` too, right? It should, and should be documented as such.

File: tests/fixtures/typescript-types (hunk 1 above)
Comment (ExE-Boss) [reply tests/fixtures/typescript-types]: Way ahead of you:
https://github.com/EB-Forks/jsdoctypeparser/blob/9329a0bbca7eea5a78ee999e5cdd5acd3c63cd58/tests/test_parsing.js#L494-501

https://github.com/EB-Forks/jsdoctypeparser/blob/9329a0bbca7eea5a78ee999e5cdd5acd3c63cd58/tests/test_parsing.js#L503-512

https://github.com/EB-Forks/jsdoctypeparser/blob/9329a0bbca7eea5a78ee999e5cdd5acd3c63cd58/tests/test_parsing.js#L526-540

**Comment by ExE-Boss** [comment]
Looks like I completely forgot about variadic tuples in the initial implementation.

**Comment by sandersn** [comment]
Can you elaborate on how suffix-dots fails in Typescript? If I remember right, it currently is part of the `function` syntax, not the type syntax. That's how it's specified in Closure I believe, and definitely how it's used in Closure code.

I had it part of the type syntax originally, but I believe somebody filed a bug on it. It also made parsing more complicated in an ad-hoc way.

**Comment by ExE-Boss** [comment]
It&nbsp;means that&nbsp;`[A, B...]` gets parsed as&nbsp;`[A, B, ...any]` instead&nbsp;of&nbsp;`[A, ...B]` by&nbsp;**TypeScript**’s JSDoc parser.

The same also applies to&nbsp;rest&nbsp;parameters within the&nbsp;`function(A, B...)`&nbsp;type.

**Pull request merged**